csharp/ql/src/experimental/CWE-295/InsecureCertificateValidationQuery.qll

```diff
@@ -0,0 +1,152 @@
+ */
+abstract private class Sanitizer extends DataFlow::ExprNode { }
+
+/**
+ * Provides an abstract base class for properties related to certificate validation.
+ */
+abstract private class CertificateValidationProperty extends Property { }
+
+/**
+ * Represents the `ServerCertificateValidationCallback` property of the `ServicePointManager` class.
+ */
+private class ServicePointManagerServerCertificateValidationCallback extends CertificateValidationProperty
+{
+  ServicePointManagerServerCertificateValidationCallback() {
+    this.getDeclaringType().hasFullyQualifiedName("System.Net", "ServicePointManager") and
+    this.hasName("ServerCertificateValidationCallback")
+  }
+}
+
+/**
+ * Represents the `ServerCertificateCustomValidationCallback` property of the `HttpClientHandler` class.
+ */
+private class HttpClientHandlerServerCertificateCustomValidationCallback extends CertificateValidationProperty
+{
+  HttpClientHandlerServerCertificateCustomValidationCallback() {
+    this.getDeclaringType().hasFullyQualifiedName("System.Net.Http", "HttpClientHandler") and
+    this.hasName("ServerCertificateCustomValidationCallback")
+  }
+}
+
+/**
+ * Represents the `ServerCertificateValidationCallback` property of the `HttpWebRequest` class.
+ */
+private class HttpWebRequestServerCertificateValidationCallback extends CertificateValidationProperty
+{
+  HttpWebRequestServerCertificateValidationCallback() {
+    this.getDeclaringType().hasFullyQualifiedName("System.Net", "HttpWebRequest") and
+    this.hasName("ServerCertificateValidationCallback")
+  }
+}
+
+/**
+ * Represents the creation of an `SslStream` object.
+ */
+private class SslStreamCreation extends ObjectCreation {
+  SslStreamCreation() {
+    this.getObjectType().hasFullyQualifiedName("System.Net.Security", "SslStream")
+  }
+
+  /**
+   * Gets the expression used as the server certificate validation callback argument
+   * in the creation of the `SslStream` object.
+   */
+  Expr getServerCertificateValidationCallback() { result = this.getArgument(2) }
+}
+
+/**
+ * Holds if `c` always returns `true`.
+ */
+private predicate alwaysReturnsTrue(Callable c) {
+  forex(ReturnStmt rs | rs.getEnclosingCallable() = c |
+    rs.getExpr().(BoolLiteral).getBoolValue() = true
+  )
+  or
+  c.getBody().(BoolLiteral).getBoolValue() = true
+}
+
+/**
+ * Gets the actual callable object referred to by expression `e`.
+ * This can be a direct reference to a callable, a delegate creation, or an anonymous function.
+ */
+private Callable getActualCallable(Expr e) {
+  exists(Expr dcArg | dcArg = e.(DelegateCreation).getArgument() |
+    result = dcArg.(CallableAccess).getTarget() or
+    result = dcArg.(AnonymousFunctionExpr)
+  )
+  or
+  result = e
+}
+
+private predicate ignoreCertificateValidation(Guard guard, AbstractValue v) {
+  guard =
+    any(PropertyAccess access |
+      access.getProperty().hasFullyQualifiedName("", "Settings", "IgnoreCertificateValidation") and
```

Comment:
@intrigus-lgtm : This is somewhat similar to the Java implementation of *flags* that you mentioned. There doesn't exist similar library functionality for C# and thus we need to implement it here.
This is hardcoded specifically to the test *flag* `Settings.IngoreCertificateValidation` just to illustrate, how this can be achieved. This code needs to be modified to the flags/conditions that fits the purposes of certificate validation.